Add faulty tensorpipe implementation

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61907 Remove faulty process group code
* **#61421 Add faulty tensorpipe implementation**

This PR adds the faulty tensorpipe agent implementation and replaces all faulty process group agent tests with it. The faulty tensorpipe agent code is very similar to that of faulty process group agent. It allows the user to fail or delay certain types of rpc messages, which is used in the faulty agent tests. These changes are needed to deprecate the process group rpc backend.

Summary of changes:
- Add faulty tensorpipe agent class
- Update tensorpipe pipeWrite function to allow to be overwritten and add delay
- Update test backend registry and faulty agent tests to use the FAULTY_TENSORPIPE backend.

This effects all faulty agent tests, here a few of them as sample commands:
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_verify_backend_options`
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_no_faulty_messages`
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_builtin_remote_message_dropped_timeout`

Differential Revision: [D29773739](https://our.internmc.facebook.com/intern/diff/D29773739)

H-Huang committed:
[draft] faulty tensorpipe changes

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] faulty tensorpipe changes"




[ghstack-poisoned]

H-Huang committed:
Update on "[draft] faulty tensorpipe changes"


sample test commands:
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_verify_backend_options
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_no_faulty_messages
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_builtin_remote_message_dropped_timeout

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] faulty tensorpipe changes"


sample test commands:
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_verify_backend_options
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_no_faulty_messages
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_builtin_remote_message_dropped_timeout

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] faulty tensorpipe changes"


sample test commands:
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_verify_backend_options
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_no_faulty_messages
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_builtin_remote_message_dropped_timeout

[ghstack-poisoned]

H-Huang committed:
Update on "[draft] faulty tensorpipe changes"


sample test commands:
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_verify_backend_options
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_no_faulty_messages
pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_builtin_remote_message_dropped_timeout

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29773739).

H-Huang committed:
Update on "Add faulty tensorpipe implementation"


This PR adds the faulty tensorpipe agent implementation and replaces all faulty process group agent tests with it. The faulty tensorpipe agent code is very similar to that of faulty process group agent. It allows the user to fail or delay certain types of rpc messages, which is used in the faulty agent tests. These changes are needed to deprecate the process group rpc backend.

Summary of changes:
- Add faulty tensorpipe agent class
- Update tensorpipe pipeWrite function to allow to be overwritten and add delay
- Update test backend registry and faulty agent tests to use the FAULTY_TENSORPIPE backend.

This effects all faulty agent tests, here a few of them as sample commands:
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_verify_backend_options`
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_no_faulty_messages`
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_builtin_remote_message_dropped_timeout`

Differential Revision: [D29773739](https://our.internmc.facebook.com/intern/diff/D29773739)

[ghstack-poisoned]

mrshenli commented on tools/build_variables.bzl:
```diff
@@ -383,6 +383,7 @@ libtorch_distributed_extra_sources = [
     "torch/csrc/distributed/rpc/tensorpipe_agent.cpp",
     "torch/csrc/distributed/rpc/tensorpipe_utils.cpp",
     "torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp",
```
Shall we remove `faulty_process_group_agent` and `process_group_agent` in followup PRs? Thanks!

H-Huang replied:
Yep! Will do that and update code comments

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -220,6 +220,16 @@ class TORCH_API TensorPipeAgent : public RpcAgent {
 
+ protected:
+  // TensorPipe write function that could be used to write response
+  // messages by server, and write request messages by client.
```
Let's add a comment to explain that this is under `protected` because we need to override it in FaultyTensorPipeAgent.

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_tensorpipe_agent.cpp:
```diff
@@ -0,0 +1,151 @@
+    std::this_thread::sleep_for(std::chrono::milliseconds(
+        static_cast<int>(msgDelay * kSecToMsConversion)));
+  }
+  TensorPipeAgent::pipeWrite(pipe, rpcMessage, std::move(devices), streams, fn);
```
Would I be correct if I assume the main difference is that we replaced `FaultyProcessGroupAgent::enqueueSend` with `FaultyTensorPipeAgent::pipeWrite`?

H-Huang replied:
Yes thats the main difference, `FaultyProcessGroupAgent::enqueueSend` as well as `FaultyProcessGroupAgent::sendToSelf`

mrshenli reviewed (approved):
LGTM!

H-Huang committed:
Update on "Add faulty tensorpipe implementation"


This PR adds the faulty tensorpipe agent implementation and replaces all faulty process group agent tests with it. The faulty tensorpipe agent code is very similar to that of faulty process group agent. It allows the user to fail or delay certain types of rpc messages, which is used in the faulty agent tests. These changes are needed to deprecate the process group rpc backend.

Summary of changes:
- Add faulty tensorpipe agent class
- Update tensorpipe pipeWrite function to allow to be overwritten and add delay
- Update test backend registry and faulty agent tests to use the FAULTY_TENSORPIPE backend.

This effects all faulty agent tests, here a few of them as sample commands:
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_verify_backend_options`
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_no_faulty_messages`
`pytest test/distributed/rpc/test_faulty_agent.py -vs -k test_builtin_remote_message_dropped_timeout`

Differential Revision: [D29773739](https://our.internmc.facebook.com/intern/diff/D29773739)

[ghstack-poisoned]
